shell/shell.js

```diff
@@ -16,22 +16,102 @@
+  this.isMac = Services.appinfo.OS === 'Darwin';
+  this.keys = {
+    i: 73,
+    r: 114,
+    f5: 116
+  };
+}
+Shortcuts.prototype.reloadPage = function(event) {
+  if (this.isMac) {
+    // `Cmd + R` or `F5`.
+    return (event.metaKey && event.keyCode === this.keys.r) || event.keyCode === this.keys.f5;
+  }
+  // `Ctrl + Shift + I`.
+  return event.ctrlKey && event.shiftKey && event.keyCode === this.keys.i;
+};
+Shortcuts.prototype.toggleDevTools = function(event) {
+  // dump(`openDevTools ${this.isMac} ${event.metaKey} ${event.altKey} ${event.keyCode === this.keys.i}\n`);
 
-  browser.loadURI(url, null, null);
-  // dump instead of console.log to write to stdout for tests.
-  dump(`opened ${url} in new window\n`);
-  Runtime.openDevTools(browser);
+  if (this.isMac) {
+    // `Cmd + Alt + I`.
+    return event.metaKey && event.altKey && event.keyCode === this.keys.i;
+  }
+  // `Ctrl + Shift + I`.
+  return event.ctrlKey && event.shiftKey && event.keyCode === this.keys.i;
+};
+Shortcuts.prototype.openDevTools = Shortcuts.prototype.toggleDevTools;
+Shortcuts.prototype.closeDevTools = Shortcuts.prototype.toggleDevTools;
 
-  browser.addEventListener('keydown', event => {
-    // Reload the web page when the F5 key is pressed.
-    if (event.keyCode && event.keyCode === 116 && !event.shiftKey &&
-        !event.altKey && !event.ctrlKey && !event.metaKey) {
-      browser.reload();
-    }
-  }, false, true);
+window.addEventListener('load', () => {
+  UI.init();
+}, { once: true });
 
-}, false);
+window.addEventListener('unload', () => {
+  UI.destroy();
+}, { once: true });
+
+const UI = {
+  init: () => {
+    const browser = document.getElementById('content');
+    const url = window.arguments[0];
+    const shortcuts = new Shortcuts();
+    let toolsWindow;
+
+    // Focus the browser window when the application is opened.
+    browser.focus();
+
+    browser.loadURI(url, null, null);
+
+    // dump instead of console.log to write to stdout for tests.
+    dump(`opened ${url} in new window\n`);
+
+    const onToolsKeydown = event => {
+      // dump(`onToolsKeydown\n`);
+      if (shortcuts.toggleDevTools(event)) {
+        // dump(`onToolsKeydown [shortcut OK]\n`);
+        Runtime.toggleDevTools(browser);
+      }
+    };
+
+    const onToolsUnload = () => {
+      // dump(`onToolsUnload\n`);
+      toolsWindow.removeEventListener('keydown', onToolsKeydown);
+    };
+
+    const onToolsLoad = () => {
+      // dump(`onToolsLoad\n`);
+      toolsWindow.addEventListener('keydown', onToolsKeydown);
+      toolsWindow.removeEventListener('unload', onToolsUnload);
+    };
+
+    browser.addEventListener('keydown', event => {
+      // Reload the web page when the `F5` key (or `Command+R` on Mac) is pressed.
+      if (shortcuts.reloadPage(event)) {
+        browser.reload();
+      }
+
+      if (shortcuts.toggleDevTools(event)) {
+        if (toolsWindow) {
+          const openedDevTools = Runtime.toggleDevTools(browser);
```

Comment:
@cvan I answered this over in https://github.com/mozilla/qbrt/pull/56#issuecomment-307225594.